src/Neo/Persistence/DataCache.cs

```diff
@@ -126,27 +126,33 @@ public void Add(StorageKey key, StorageItem value)
         public virtual void Commit()
         {
             LinkedList<StorageKey> deletedItem = new();
-            foreach (Trackable trackable in GetChangeSet())
-                switch (trackable.State)
+            lock (dictionary)
+            {
+                foreach (var key in changeSet)
                 {
-                    case TrackState.Added:
-                        AddInternal(trackable.Key, trackable.Item);
-                        trackable.State = TrackState.None;
-                        break;
-                    case TrackState.Changed:
-                        UpdateInternal(trackable.Key, trackable.Item);
-                        trackable.State = TrackState.None;
-                        break;
-                    case TrackState.Deleted:
-                        DeleteInternal(trackable.Key);
-                        deletedItem.AddFirst(trackable.Key);
-                        break;
+                    var trackable = dictionary[key];
+                    switch (trackable.State)
+                    {
+                        case TrackState.Added:
+                            AddInternal(key, trackable.Item);
+                            trackable.State = TrackState.None;
+                            break;
+                        case TrackState.Changed:
+                            UpdateInternal(key, trackable.Item);
+                            trackable.State = TrackState.None;
+                            break;
+                        case TrackState.Deleted:
+                            DeleteInternal(key);
+                            deletedItem.AddFirst(key);
```

Comment:
Why we don't Delete directly here, is not possible to repeat a key in `changeSet` ?